### PR TITLE
Feature/language

### LINKED ISSUE #13777: [feature] Add cstd to compiler settings
**Background**
We encountered an issue with the GCC toolchain (at least GCC 9) pulling in BSD and SVr4 extensions if no C/C++ dialect via compiler flag _-std=_ has been explicitly defined. This is unwanted on many embedded targets.

We found out that in contrast to the legacy [Conan CMake build helper](https://docs.conan.io/en/1.59/reference/build_helpers/cmake.html) the new [generate()](https://docs.conan.io/en/1.59/migrating_to_2.0/recipes.html#the-generate-method) automatically translates the infomation in setting _compiler.cppstd_ to a toolchain specific flag and appends it to the compiler call. This works pretty well for C++ files, but as many of our packages contain C files or are written exclusively in C, there is no similar feature for C in Conan. Using the same flag as for C++ is no option as this will create warnings for C files.

**Proposal**
- Add _cstd_ to settings _compiler_ similar to _cppstd_.
- Translate the information in _compiler.cstd_ to a compiler specific flag (e.g. for GCC _compiler.cstd=11_ translates to _-std=c11_).
- If using the [generate()](https://docs.conan.io/en/1.59/migrating_to_2.0/recipes.html#the-generate-method) approach add this compile flag to the compiler call.

### Have you read the CONTRIBUTING guide?

- [X] I've read the CONTRIBUTING guide

### PR DESCRIPTION
**Pull request by memsharded**
Changelog: Feature: New ``compiler.cstd`` setting for C standard
Changelog: Feature: Implemented ``compatibility.py`` default compatibility for different C standards
Changelog: Feature: Implement ``check_min_cstd``, ``check_max_cstd``, ``valid_max_cstd``, ``valid_min_cstd``, ``supported_cstd`` tools
Changelog: Feature: New ``languages = "C", "C++"`` class attribute to further automate settings management
Docs: https://github.com/conan-io/docs/pull/3757

Close https://github.com/conan-io/conan/issues/13777

### CODE REVIEW

**Review by AbrilRBS — APPROVED** [review]

File: conans/model/conan_file.py [hunk 1]
```diff
@@ -48,6 +48,7 @@ class ConanFile:
     default_options = None
     default_build_options = None
     package_type = None
+    languages = []
```
Comment (AbrilRBS) [inline conans/model/conan_file.py:50]: Can't forget to serialize this, and make it available in the conanfile interface

**Comment by jcar87** [comment]
This is great!!

I wonder if we could address the usage of the `stdcpp_library()` helper.
Some recipes build libraries that have a pure "C" interface but encapsulate compiled C++ code. Downstream pure C consumers linking the _static_ library will need to tell the linker that the standard C++ library needs to be linked as well:

Usually handled like this:
```
    def package_info(self):
        self.cpp_info.libs = ["zimg"]
        if not self.options.shared:
            libcxx = stdcpp_library(self)
            if libcxx:
                self.cpp_info.system_libs.append(libcxx)
```

This is not entirely robust, the downstream consumer may choose to use the _static_ C++ library, which is valid - but Conan would add a conflicting flag.

Perhaps we can take advantage of this feature to make this logic redundant but also robust?

**Review by AbrilRBS — COMMENTED** [review]
We still need to decide if the profile detectino should fill a value for the cstd by default

**Comment by memsharded** [comment]
We have a problem with current approach:

- If we define ``compiler.cstd=11`` in profile
- Use a conanfile.txt (no ``languages = "C", ...`` specification)
- The ``cstd`` is always auto-removed
- It is never passed to the toolchain (see failing test in https://github.com/conan-io/conan/pull/16028/commits/2a870159489b287b1b4bb8b1f1183642ed01d2d6)

**Comment by memsharded** [comment]
> We have a problem with current approach:

> - If we define compiler.cstd=11 in profile
> - Use a conanfile.txt (no languages = "C", ... specification)
> - The cstd is always auto-removed
> - it is never passed to the toolchain (see failing test in https://github.com/conan-io/conan/commit/2a870159489b287b1b4bb8b1f1183642ed01d2d6)

Users wanting to support local dev must use a ``conanfile.py``, not a ``conanfile.txt``?

**Comment by memsharded** [comment]
> I wonder if we could address the usage of the stdcpp_library() helper.

@jcar87 I am trying to check what would be the more idiomatic CMake way of doing this, but I can't find anything easy. Is there any property of targets that could be used for this? If that is the case, we could do it in ``CMakeDeps``. If that is not the case, we might want to do it in ``CMakeToolchain``, but that might have limitations like not usable for ``cmake-conan`` or not possible to use it per-target, while ``CMakeDeps`` will apply only to targets linking certain dependencies.

**Review by czoido — APPROVED** [review]

File: conan/tools/build/cstd.py [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+
+
+def _apple_clang_supported_cstd(version):
+    # TODO: Per-version support
+    return ["99", "gnu99", "11", "gnu11", "17", "gnu17", "23", "gnu23"]
+
+
+def _gcc_supported_cstd(version):
+    # TODO: Per-version support
+    return ["99", "gnu99", "11", "gnu11", "17", "gnu17", "23", "gnu23"]
```
Comment (AbrilRBS) [inline conan/tools/build/cstd.py]: I generated this link to check the versions: https://godbolt.org/z/4nTYcfMT5

```suggestion
    if version < "4.7":
        return ["99", "gnu99"]
    if version < "8":
        return ["99", "gnu99", "11", "gnu11"]
    if version < "14":
        return ["99", "gnu99", "11", "gnu11", "17", "gnu17"]
    return ["99", "gnu99", "11", "gnu11", "17", "gnu17", "23", "gnu23"]
```

The `cppstd` file has a 3.4 verison check to renturn an empty version but I have no access to such an old release

Comment (memsharded) [reply conan/tools/build/cstd.py]: Nice, many thanks!!

File: conan/tools/build/cstd.py [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+
+
+def _msvc_supported_cstd(version):
+    # TODO: Per-version support
+    return ["11", "17"]
+
+
+def _clang_supported_cstd(version):
+    # TODO: Per-version support
+    return ["99", "gnu99", "11", "gnu11", "17", "gnu17", "23", "gnu23"]
```
Comment (AbrilRBS) [inline conan/tools/build/cstd.py]: From https://godbolt.org/z/r5cq7s651

```suggestion
    if version < "3":
        return ["99", "gnu99"]
    if version < "6":
        return ["99", "gnu99", "11", "gnu11"]
    if version < "18":
        return ["99", "gnu99", "11", "gnu11", "17", "gnu17"]
    return ["99", "gnu99", "11", "gnu11", "17", "gnu17", "23", "gnu23"]
```

File: conan/tools/build/cstd.py [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+
+
+def _gcc_supported_cstd(version):
+    # TODO: Per-version support
+    return ["99", "gnu99", "11", "gnu11", "17", "gnu17", "23", "gnu23"]
+
+
+def _msvc_supported_cstd(version):
+    # TODO: Per-version support
+    return ["11", "17"]
```
Comment (AbrilRBS) [inline conan/tools/build/cstd.py]: From https://learn.microsoft.com/en-us/cpp/build/reference/std-specify-language-standard-version?view=msvc-170#c-standards-support-1

```suggestion
    if version < "192":
        return []
    return ["11", "17"]
```

**Review by jcar87 — APPROVED** [review]

**Pull request merged**